### ISSUE TITLE
Add a flag to skip syncHostData() in goToDefinition()

### ISSUE THREAD

**Issue by joshribakoff** [microsoft/TypeScript#32406]
## Search Terms

goToDefinition go to definition 

## Suggestion

I'm writing a static analyzer which makes use of `goToDefinition`. I need to add a flag to bypass cache invalidation. In a large project, `goToDefinition` takes 300ms. When writing a static analyzer, I load all files into memory & crawl their dependencies. I know when my static analyzer has modified code in memory, so I need a flag to tell TS when to sync the host data, and avoid the 300ms check each time which makes the static analysis so slow it is no longer feasible.  With this flag, analysis completes in minutes instead of days or weeks.

## Use Cases

Static analysis

Obviously this would not be viable in the editor

## Examples

Jumping to definition takes around 300ms in Typescript, when using the compiler API. I found that if you modify `./node_modules/typescript/lib/typescript.js` to add a skip flag, it can jump to definition in under `.5ms`:

```
/// Goto definition
function getDefinitionAtPosition(fileName, position, skip) {
    synchronizeHostData(skip);
    // ....
}

function synchronizeHostData(skip= false) {
  if(skip) return;
  // ....
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by sheetalkamat** [comment]
You can implement `getProjectVersion` on `LanguageServiceHost` that can make sure that program is not updated if version has not changed.

**Comment by joshribakoff** [comment]
@sheetalkamat thanks, but I am using ts-morph (https://github.com/dsherret/ts-morph)... so I guess I would need to open a new issue there for them to implement it... I am not directly implementing a language service host in my project, as that would be a lot of boilerplate for me to do in my project directly.
